server/conf/nginx/nginx.conf

```diff
@@ -85,6 +85,7 @@ http {
             proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
             proxy_set_header Host $host;
             proxy_redirect off;
+            proxy_set_header X-Forwarded-Proto https;
```

Comment:
Django v4 has some extra security when making POST and PUT requests to a backend where it checks the Origin/Referrer header and only allows requests to go through when that matches with the Host header. See reference [here](https://docs.djangoproject.com/en/4.2/ref/settings/#csrf-trusted-origins). This was causing an error where POST requests from the portal were not working (on cep.test at least) with the following error: 

`Forbidden (Origin checking failed - https://cep.test does not match any trusted origins.)`

I figured out a couple solutions for this: 

1- One way to fix this issue is to add cep.test as a a trusted origin using the CSRF_TRUSTED_ORIGINS django setting which would look something like this: 

`CSRF_TRUSTED_ORIGINS = ['https://cep.test'] `

I am not entirely sure if we would need to add origins for all the portals here. 

2- Another solution (which I implemented) is to add an additional configuration to the nginx config file where we explicitly set the request header to https which allows django to interpret the incoming request as secure. This method seems to work and the only issue I ran into here is that for the cep.test client in tapis the callback url is `http://cep.test/auth/tapis/callback/ `and with this additional config the redirect url sent in is `https://cep.test/auth/tapis/callback/` so there is a protocol mismatch which causes an invalid redirect url error to occur. This can be fixed by updating the cep.test client callback url to use https and since there aren't any other clients that use an http for callback urls I don't believe we should run into this issue again. 

Functionality wise everything seems to be working with this added config but I am not sure if there is anything I am missing or if there's another issue anyone can foresee by adding this.

Reply:
This is only applicable for local based on our configuration.

In actual deployment, it uses [uwsgi_pass](https://github.com/TACC/Camino/blob/1de9c55bbc904f3367e9679fe021dc2ee5c3b8dd/conf/nginx/templates/default.conf.template#L75) and [HTTP_X_FORWARDED_PROTO](https://github.com/TACC/Camino/blob/main/conf/uwsgi/uwsgi_params#L21C14-L21C36) in uswgi_param. So, the django strict middleware checks in 4.x should allow it through.

One minor nit: use $scheme instead of https ? 

CSRF_TRUSTED_ORIGINS is for cross domain requests, which we are not doing. It is a workaround that django allows and works for now in this scenario.

Reply:
Thanks for that info. So for just local testing would the better solution be to remove this extra nginx config and add in `CSRF_TRUSTED_ORIGINS = ['https://cep.test']` in django settings?

Reply:
CSRF_TRUSTED_ORIGINS only for local settings may be ok for now.